File: packages/protocol-ts/tests/registries.test.ts
/**
 * Tests de los registries del SDK.
 *
 * Verifica que cada registry tiene los counts esperados, que cada entry
 * tiene shape válido, y que las funciones de acceso (getX/allX) funcionan
 * coherentemente.
 *
 * Estos tests cazan regresiones del tipo "alguien borró un behavior sin
 * actualizar el corpus" o "alguien cambió el shape de RecipeManifest".
 */

import { describe, it, expect } from 'vitest';
import {
  allFieldTypes,
  getFieldType,
  FIELD_TYPE_IDS,
  type FieldTypeDefinition,
} from '../src/registries/field-types';
import {
  allActions,
  getAction,
  ACTION_IDS,
  type ActionDefinition,
} from '../src/registries/actions';
import {
  allBehaviors,
  getBehavior,
  getBehaviorsByType,
  suggestBehavior,
  type BehaviorDefinition,
} from '../src/registries/behaviors';
import {
  RECIPE_REGISTRY,
  getRecipeManifest,
  allRecipes,
  allRecipesByKind,
  type RecipeManifest,
} from '../src/registries/recipes';
import {
  SLOT_ACCEPT_KIND_META,
  getSlotAcceptKindOptions,
  formatSlotAccepts,
  getAvailableAppearanceProps,
} from '../src/registries/slot-kinds';
import type { SlotAcceptKind } from '../src/types';

// ── Field types ──────────────────────────────────────────────────────

describe('field-type-registry', () => {
  const types = allFieldTypes();

  it('contiene 9 types base', () => {
    // KRO-75: añadido cardRef para resolver drift con Studio.
    expect(types).toHaveLength(9);
  });

  it('incluye los IDs core', () => {
    const ids = types.map(t => t.id);
    expect(ids).toEqual(
      expect.arrayContaining([
        'text', 'textarea', 'number', 'select', 'image',
        'array<string>', 'array<number>', 'array<image>',
        'cardRef',
      ]),
    );
  });

  it('cada entry tiene shape válido', () => {
    types.forEach(t => {
      // KRO-75: el regex acepta camelCase (cardRef) además de lowercase
      // + delimitadores de tipo paramétrico.
      expect(t.id).toMatch(/^[a-zA-Z<>]+$/);
      expect(t.displayName).toBeTruthy();
      expect(t.description.length).toBeGreaterThan(20);
      expect(['scalar', 'array']).toContain(t.cardinality);
      if (t.cardinality === 'array') {
        expect(t.elementType).toBeTruthy();
      }
    });
  });

  it('FIELD_TYPE_IDS matchea los ids del catálogo', () => {
    expect([...FIELD_TYPE_IDS].sort()).toEqual(types.map(t => t.id).sort());
  });

  it('getFieldType devuelve la entry correcta', () => {
    const text = getFieldType('text');
    expect(text).toBeDefined();
    expect(text!.cardinality).toBe('scalar');

    const arrImg = getFieldType('array<image>');
    expect(arrImg!.cardinality).toBe('array');
    expect(arrImg!.elementType).toBe('image');
  });

  it('getFieldType devuelve undefined para id inexistente', () => {
    expect(getFieldType('unknown')).toBeUndefined();
  });
});

// ── Actions ──────────────────────────────────────────────────────────

describe('action-registry', () => {
  const actions = allActions();

  it('contiene 5 actions', () => {
    expect(actions).toHaveLength(5);
  });

  it('incluye los IDs core', () => {
    const ids = actions.map(a => a.id);
    expect(ids).toEqual(
      expect.arrayContaining([
        'none', 'navigate_to_detail', 'modal', 'expand_inline', 'external_link',
      ]),
    );
  });

  it('cada action tiene shape válido', () => {
    actions.forEach(a => {
      expect(a.id).toMatch(/^[a-z_]+$/);
      expect(a.displayName).toBeTruthy();
      expect(a.description).toBeTruthy();
      expect(['static', 'push', 'modal', 'inline', 'external']).toContain(a.transition);
    });
  });

  it('constraint flags coherentes con transition', () => {
    const byId = (id: string) => actions.find(a => a.id === id)!;
    expect(byId('navigate_to_detail').requiresTargetRecipe).toBe(true);
    expect(byId('navigate_to_detail').targetRecipeKind).toBe('detail');
    expect(byId('modal').requiresTargetRecipe).toBe(true);
    expect(byId('expand_inline').requiresExpandRecipe).toBe(true);
    expect(byId('external_link').requiresLinkField).toBe(true);
    expect(byId('none').requiresTargetRecipe).toBeFalsy();
  });

  it('ACTION_IDS matchea los ids', () => {
    expect([...ACTION_IDS].sort()).toEqual(actions.map(a => a.id).sort());
  });

  it('getAction devuelve la entry correcta', () => {
    expect(getAction('none')?.transition).toBe('static');
    expect(getAction('unknown')).toBeUndefined();
  });
});

// ── Behaviors ────────────────────────────────────────────────────────

describe('behavior-registry', () => {
  const behaviors = allBehaviors();

  it('contiene 26 behaviors', () => {
    expect(behaviors).toHaveLength(26);
  });

  it('IDs únicos (sin duplicados)', () => {
    const ids = behaviors.map(b => b.id);
    expect(new Set(ids).size).toBe(ids.length);
  });

  it('cada behavior tiene shape válido', () => {
    behaviors.forEach(b => {
      expect(b.id).toMatch(/^[a-z_]+$/);
      expect(b.displayName).toBeTruthy();
      expect(b.description.length).toBeGreaterThan(20);
      expect(b.applicableTypes.length).toBeGreaterThan(0);
      // renderAsSlotKind es opcional, pero si está, debe estar en el meta
      if (b.renderAsSlotKind) {
        expect(SLOT_ACCEPT_KIND_META).toHaveProperty(b.renderAsSlotKind);
      }
    });
  });

  it('applicableTypes solo contiene types base válidos', () => {
    behaviors.forEach(b => {
      b.applicableTypes.forEach(t => {
        expect(FIELD_TYPE_IDS).toContain(t);
      });
    });
  });

  it('getBehavior y getBehaviorsByType funcionan', () => {
    expect(getBehavior('color_hex')?.renderAsSlotKind).toBe('color');
    expect(getBehavior(undefined)).toBeUndefined();
    expect(getBehavior(null)).toBeUndefined();
    expect(getBehavior('unknown')).toBeUndefined();

    const textBehaviors = getBehaviorsByType('text');
    expect(textBehaviors.length).toBeGreaterThan(0);
    textBehaviors.forEach(b => expect(b.applicableTypes).toContain('text'));
  });

  it('suggestBehavior matchea patterns conocidos', () => {
    expect(suggestBehavior('email', 'text')).toBe('email');
    expect(suggestBehavior('year', 'number')).toBe('year');
    expect(suggestBehavior('color', 'text')).toBe('color_hex');
    // Tipo incompatíble
    expect(suggestBehavior('email', 'number')).toBeUndefined();
    // Sin pattern
    expect(suggestBehavior('foobar', 'text')).toBeUndefined();
  });
});

// ── Recipes ──────────────────────────────────────────────────────────

describe('recipe-registry', () => {
  const recipes = allRecipes();

  it('contiene 8 recipes', () => {
    expect(recipes).toHaveLength(8);
  });

  it('distribución por kind: 3 list + 3 detail + 2 expand', () => {
    const byKind = (k: string) => recipes.filter(r => r.kind === k);
    expect(byKind('list')).toHaveLength(3);
    expect(byKind('detail')).toHaveLength(3);
    expect(byKind('expand')).toHaveLength(2);
  });

  it('IDs únicos', () => {
    const ids = recipes.map(r => r.id);
    expect(new Set(ids).size).toBe(ids.length);
  });

  it('cada recipe tiene shape válido', () => {
    recipes.forEach(r => {
      expect(r.id).toMatch(/^[a-z_]+$/);
      expect(r.displayName).toBeTruthy();
      expect(r.description).toBeTruthy();
      expect(['list', 'detail', 'expand']).toContain(r.kind);
      expect(r.slots.length).toBeGreaterThan(0);
    });
  });

  it('cada slot tiene shape válido + accepts existentes en SLOT_ACCEPT_KIND_META', () => {
    recipes.forEach(r => {
      const slotIds = new Set<string>();
      r.slots.forEach(s => {
        expect(s.id).toMatch(/^[a-z_]+$/);
        expect(slotIds.has(s.id)).toBe(false); // sin duplicados en la misma recipe
        slotIds.add(s.id);
        expect(s.label).toBeTruthy();
        expect(['single', 'composable']).toContain(s.kind);
        expect(s.accepts.length).toBeGreaterThan(0);
        s.accepts.forEach(k => {
          expect(SLOT_ACCEPT_KIND_META).toHaveProperty(k);
        });
      });
    });
  });

  it('getRecipeManifest devuelve correctamente', () => {
    const ca = getRecipeManifest('compact_avatar');
    expect(ca).toBeDefined();
    expect(ca!.kind).toBe('list');

    // ID que NO existe en RECIPE_REGISTRY pero SÍ en el union → undefined
    // (no hay; todos los IDs del union existen en el registry actualmente).
  });

  it('RECIPE_REGISTRY tiene una entry por cada recipe', () => {
    expect(Object.keys(RECIPE_REGISTRY).sort()).toEqual(recipes.map(r => r.id).sort());
  });

  // KRO-76 — Helper allRecipesByKind
  describe('allRecipesByKind', () => {
    it('list → 3 recipes', () => {
      const list = allRecipesByKind('list');
      expect(list.length).toBeGreaterThanOrEqual(1);
      list.forEach(r => expect(r.kind).toBe('list'));
    });

    it('detail → 3 recipes', () => {
      const detail = allRecipesByKind('detail');
      expect(detail.length).toBeGreaterThanOrEqual(1);
      detail.forEach(r => expect(r.kind).toBe('detail'));
    });

    it('expand → 2 recipes', () => {
      const expand = allRecipesByKind('expand');
      expect(expand.length).toBeGreaterThanOrEqual(1);
      expand.forEach(r => expect(r.kind).toBe('expand'));
    });

    it('sum por kind === total allRecipes()', () => {
      const total = allRecipes().length;
      const sum = allRecipesByKind('list').length
                + allRecipesByKind('detail').length
                + allRecipesByKind('expand').length;
      expect(sum).toBe(total);
    });

    it('mismas referencias que el registry (no copia)', () => {
      const list = allRecipesByKind('list');
      list.forEach(r => {
        expect(RECIPE_REGISTRY[r.id]).toBe(r);
      });
    });
  });
});

// ── Slot kinds (meta) ────────────────────────────────────────────────

describe('slot-kinds (SLOT_ACCEPT_KIND_META)', () => {
  const kinds = Object.keys(SLOT_ACCEPT_KIND_META) as SlotAcceptKind[];

  it('contiene 14 kinds', () => {
    expect(kinds).toHaveLength(14);
  });

  it('cada kind tiene label + description no-vacíos', () => {
    kinds.forEach(k => {
      const meta = SLOT_ACCEPT_KIND_META[k];
      expect(meta.label).toBeTruthy();
      expect(meta.description.length).toBeGreaterThan(10);
    });
  });

  it('getSlotAcceptKindOptions devuelve array con el shape correcto', () => {
    const opts = getSlotAcceptKindOptions();
    expect(opts.length).toBe(kinds.length);
    opts.forEach(o => {
      expect(o.id).toBeTruthy();
      expect(o.label).toBeTruthy();
      expect(o.description).toBeTruthy();
    });
  });

  it('formatSlotAccepts: any → "cualquiera"', () => {
    expect(formatSlotAccepts(['any'])).toBe('cualquiera');
    expect(formatSlotAccepts(['any', 'text-short'])).toBe('cualquiera');
  });

  it('formatSlotAccepts: 1 kind → su label', () => {
    expect(formatSlotAccepts(['image'])).toBe('Imagen');
  });

  it('formatSlotAccepts: N kinds → labels con " / "', () => {
    expect(formatSlotAccepts(['text-short', 'date'])).toBe('Texto / Fecha');
  });

  it('formatSlotAccepts: empty array → ""', () => {
    expect(formatSlotAccepts([])).toBe('');
  });

  it('getAvailableAppearanceProps: image-* tienen shape/aspect/imageFocus', () => {
    const props = getAvailableAppearanceProps(['image']);
    expect(props).toEqual(expect.arrayContaining(['shape', 'aspect', 'imageFocus']));
  });

  it('getAvailableAppearanceProps: text-* tienen align/weight/truncate', () => {
    const props = getAvailableAppearanceProps(['text-short']);
    expect(props).toEqual(expect.arrayContaining(['align', 'weight', 'truncate']));
  });

  it('getAvailableAppearanceProps: empty accepts → []', () => {
    expect(getAvailableAppearanceProps([])).toEqual([]);
  });

  it('getAvailableAppearanceProps: multi-accept retorna unión de props', () => {
    const props = getAvailableAppearanceProps(['image', 'text-short']);
    // Une props de image (shape, aspect, ...) con text-short (align, weight, ...)
    expect(props).toEqual(expect.arrayContaining(['shape', 'align']));
  });
});

// Helper para usar como type-check (vitest no necesita compile-time check)
type _CheckExhaustive = FieldTypeDefinition | ActionDefinition | BehaviorDefinition | RecipeManifest;
const _: _CheckExhaustive | undefined = undefined;
void _;

File: packages/core_dart/test/registries_test.dart
/// Corpus de los registries catálogo — ESPEJO de los bloques `field-type-registry`,
/// `action-registry` y `slot-kinds` de `packages/core/tests/registries.test.ts`.
///
/// Mismos counts, mismos ids, mismos flags y mismos outputs de helper que el TS.
/// Paridad por construcción: si el catálogo Dart diverge del TS, esto falla.
/// (Los bloques behavior-registry y recipe-registry se añadirán al espejar esos
/// registries en chunks siguientes.)
import 'package:test/test.dart';
import 'package:kromia_core/kromia_core.dart';

void main() {
  // ── Field types ────────────────────────────────────────────────────
  group('field-type-registry', () {
    final types = allFieldTypes();

    test('contiene 9 types base', () {
      expect(types.length, 9);
    });

    test('incluye los IDs core', () {
      final ids = types.map((t) => t.id).toList();
      expect(
        ids,
        containsAll(<String>[
          'text', 'textarea', 'number', 'select', 'image',
          'array<string>', 'array<number>', 'array<image>', 'cardRef',
        ]),
      );
    });

    test('cada entry tiene shape válido', () {
      for (final t in types) {
        expect(t.id, matches(RegExp(r'^[a-zA-Z<>]+$')));
        expect(t.displayName, isNotEmpty);
        expect(t.description.length, greaterThan(20));
        expect(<String>['scalar', 'array'], contains(t.cardinality));
        if (t.cardinality == 'array') {
          expect(t.elementType, isNotNull);
          expect(t.elementType, isNotEmpty);
        }
      }
    });

    test('fieldTypeIds matchea los ids del catálogo', () {
      final a = List<String>.of(fieldTypeIds)..sort();
      final b = types.map((t) => t.id).toList()..sort();
      expect(a, equals(b));
    });

    test('getFieldType devuelve la entry correcta', () {
      final text = getFieldType('text');
      expect(text, isNotNull);
      expect(text!.cardinality, 'scalar');

      final arrImg = getFieldType('array<image>');
      expect(arrImg!.cardinality, 'array');
      expect(arrImg.elementType, 'image');
    });

    test('getFieldType devuelve null para id inexistente', () {
      expect(getFieldType('unknown'), isNull);
    });
  });

  // ── Actions ────────────────────────────────────────────────────────
  group('action-registry', () {
    final actions = allActions();

    test('contiene 5 actions', () {
      expect(actions.length, 5);
    });

    test('incluye los IDs core', () {
      final ids = actions.map((a) => a.id).toList();
      expect(
        ids,
        containsAll(<String>[
          'none', 'navigate_to_detail', 'modal', 'expand_inline', 'external_link',
        ]),
      );
    });

    test('cada action tiene shape válido', () {
      for (final a in actions) {
        expect(a.id, matches(RegExp(r'^[a-z_]+$')));
        expect(a.displayName, isNotEmpty);
        expect(a.description, isNotEmpty);
        expect(
          <String>['static', 'push', 'modal', 'inline', 'external'],
          contains(a.transition),
        );
      }
    });

    test('constraint flags coherentes con transition', () {
      ActionDefinition byId(String id) => actions.firstWhere((a) => a.id == id);
      expect(byId('navigate_to_detail').requiresTargetRecipe, isTrue);
      expect(byId('navigate_to_detail').targetRecipeKind, 'detail');
      expect(byId('modal').requiresTargetRecipe, isTrue);
      expect(byId('expand_inline').requiresExpandRecipe, isTrue);
      expect(byId('external_link').requiresLinkField, isTrue);
      expect(byId('none').requiresTargetRecipe, isFalse);
    });

    test('actionIds matchea los ids', () {
      final a = List<String>.of(actionIds)..sort();
      final b = actions.map((a) => a.id).toList()..sort();
      expect(a, equals(b));
    });

    test('getAction devuelve la entry correcta', () {
      expect(getAction('none')?.transition, 'static');
      expect(getAction('unknown'), isNull);
    });
  });

  // ── Behaviors ──────────────────────────────────────────────────────
  group('behavior-registry', () {
    final behaviors = allBehaviors();

    test('contiene 27 behaviors', () {
      expect(behaviors.length, 27);
    });

    test('IDs únicos (sin duplicados)', () {
      final ids = behaviors.map((b) => b.id).toList();
      expect(ids.toSet().length, ids.length);
    });

    test('cada behavior tiene shape válido', () {
      for (final b in behaviors) {
        expect(b.id, matches(RegExp(r'^[a-z_]+$')));
        expect(b.displayName, isNotEmpty);
        expect(b.description.length, greaterThan(20));
        expect(b.applicableTypes.length, greaterThan(0));
        // renderAsSlotKind es opcional, pero si está, debe estar en el meta
        if (b.renderAsSlotKind != null) {
          expect(slotAcceptKindMeta.containsKey(b.renderAsSlotKind), isTrue);
        }
      }
    });

    test('applicableTypes solo contiene types base válidos', () {
      for (final b in behaviors) {
        for (final t in b.applicableTypes) {
          expect(fieldTypeIds, contains(t));
        }
      }
    });

    test('getBehavior y getBehaviorsByType funcionan', () {
      expect(getBehavior('color_hex')?.renderAsSlotKind, 'color');
      expect(getBehavior(null), isNull);
      expect(getBehavior(''), isNull);
      expect(getBehavior('unknown'), isNull);

      final textBehaviors = getBehaviorsByType('text');
      expect(textBehaviors.length, greaterThan(0));
      for (final b in textBehaviors) {
        expect(b.applicableTypes, contains('text'));
      }
    });

    test('suggestBehavior matchea patterns conocidos', () {
      expect(suggestBehavior('email', 'text'), 'email');
      expect(suggestBehavior('year', 'number'), 'year');
      expect(suggestBehavior('color', 'text'), 'color_hex');
      // Tipo incompatible
      expect(suggestBehavior('email', 'number'), isNull);
      // Sin pattern
      expect(suggestBehavior('foobar', 'text'), isNull);
    });
  });

  // ── Recipes ────────────────────────────────────────────────────────
  group('recipe-registry', () {
    final recipes = allRecipes();

    test('contiene 13 recipes (8 clásicas + 5 V5 block-native, KRO-133)', () {
      expect(recipes.length, 13);
    });

    test('distribución por kind: 7 list + 4 detail + 2 expand', () {
      List<RecipeManifest> byKind(String k) =>
          recipes.where((r) => r.kind == k).toList();
      expect(byKind('list').length, 7);
      expect(byKind('detail').length, 4);
      expect(byKind('expand').length, 2);
    });

    test('IDs únicos', () {
      final ids = recipes.map((r) => r.id).toList();
      expect(ids.toSet().length, ids.length);
    });

    test('cada recipe tiene shape válido', () {
      for (final r in recipes) {
        expect(r.id, matches(RegExp(r'^[a-z_]+$')));
        expect(r.displayName, isNotEmpty);
        expect(r.description, isNotEmpty);
        expect(<String>['list', 'detail', 'expand'], contains(r.kind));
        expect(r.slots.length, greaterThan(0));
      }
    });

    test('cada slot tiene shape válido + accepts en slotAcceptKindMeta', () {
      for (final r in recipes) {
        final slotIds = <String>{};
        for (final s in r.slots) {
          expect(s.id, matches(RegExp(r'^[a-z_]+$')));
          expect(slotIds.contains(s.id), isFalse); // sin duplicados en la misma recipe
          slotIds.add(s.id);
          expect(s.label, isNotEmpty);
          expect(<String>['single', 'composable'], contains(s.kind));
          expect(s.accepts.length, greaterThan(0));
          for (final k in s.accepts) {
            expect(slotAcceptKindMeta.containsKey(k), isTrue);
          }
        }
      }
    });

    test('getRecipeManifest devuelve correctamente', () {
      final ca = getRecipeManifest('compact_avatar');
      expect(ca, isNotNull);
      expect(ca!.kind, 'list');
    });

    test('recipeRegistry tiene una entry por cada recipe', () {
      final a = recipeRegistry.keys.toList()..sort();
      final b = recipes.map((r) => r.id).toList()..sort();
      expect(a, equals(b));
    });

    group('allRecipesByKind', () {
      test('list → 3 recipes', () {
        final list = allRecipesByKind('list');
        expect(list.length, greaterThanOrEqualTo(1));
        for (final r in list) {
          expect(r.kind, 'list');
        }
      });

      test('detail → 3 recipes', () {
        final detail = allRecipesByKind('detail');
        expect(detail.length, greaterThanOrEqualTo(1));
        for (final r in detail) {
          expect(r.kind, 'detail');
        }
      });

      test('expand → 2 recipes', () {
        final expand = allRecipesByKind('expand');
        expect(expand.length, greaterThanOrEqualTo(1));
        for (final r in expand) {
          expect(r.kind, 'expand');
        }
      });

      test('sum por kind === total allRecipes()', () {
        final total = allRecipes().length;
        final sum = allRecipesByKind('list').length +
            allRecipesByKind('detail').length +
            allRecipesByKind('expand').length;
        expect(sum, total);
      });

      test('mismas referencias que el registry (no copia)', () {
        final list = allRecipesByKind('list');
        for (final r in list) {
          expect(recipeRegistry[r.id], same(r));
        }
      });
    });
  });

  // ── Slot kinds (meta) ──────────────────────────────────────────────
  group('slot-kinds (slotAcceptKindMeta)', () {
    final kinds = slotAcceptKindMeta.keys.toList();

    test('contiene 14 kinds', () {
      expect(kinds.length, 14);
    });

    test('cada kind tiene label + description no-vacíos', () {
      for (final k in kinds) {
        final meta = slotAcceptKindMeta[k]!;
        expect(meta.label, isNotEmpty);
        expect(meta.description.length, greaterThan(10));
      }
    });

    test('getSlotAcceptKindOptions devuelve lista con el shape correcto', () {
      final opts = getSlotAcceptKindOptions();
      expect(opts.length, kinds.length);
      for (final o in opts) {
        expect(o.id, isNotEmpty);
        expect(o.label, isNotEmpty);
        expect(o.description, isNotEmpty);
      }
    });

    test('formatSlotAccepts: any → "cualquiera"', () {
      expect(formatSlotAccepts(['any']), 'cualquiera');
      expect(formatSlotAccepts(['any', 'text-short']), 'cualquiera');
    });

    test('formatSlotAccepts: 1 kind → su label', () {
      expect(formatSlotAccepts(['image']), 'Imagen');
    });

    test('formatSlotAccepts: N kinds → labels con " / "', () {
      expect(formatSlotAccepts(['text-short', 'date']), 'Texto / Fecha');
    });

    test('formatSlotAccepts: lista vacía → ""', () {
      expect(formatSlotAccepts([]), '');
    });

    test('getAvailableAppearanceProps: image → forma + objectFit + efectos (KRO-147 F3)', () {
      final props = getAvailableAppearanceProps(['image']);
      expect(props,
          containsAll(<String>['shape', 'aspect', 'objectFit', 'imageFocus', 'size', 'paddingY', 'opacity', 'shadow']));
    });

    test('getAvailableAppearanceProps: text-short → tipografía rica + color (KRO-147 F3)', () {
      final props = getAvailableAppearanceProps(['text-short']);
      expect(
          props,
          containsAll(<String>[
            'align', 'weight', 'italic', 'underline', 'textTransform', 'font',
            'lineHeight', 'tracking', 'textShadow', 'display', 'textColor', 'bgColor', 'truncate',
          ]));
    });

    test('getAvailableAppearanceProps: card-ref → shape/refSize/refColumns/refTap', () {
      final props = getAvailableAppearanceProps(['card-ref']);
      expect(props, containsAll(<String>['shape', 'refSize', 'refColumns', 'refTap']));
    });

    test('getAvailableAppearanceProps: accepts vacío → []', () {
      expect(getAvailableAppearanceProps([]), <String>[]);
    });

    test('getAvailableAppearanceProps: multi-accept retorna unión de props', () {
      final props = getAvailableAppearanceProps(['image', 'text-short']);
      expect(props, containsAll(<String>['shape', 'align']));
    });

    test('getAvailableAppearanceProps respeta el orden del catálogo (allAppearanceProps)', () {
      final props = getAvailableAppearanceProps(['image', 'text-short', 'card-ref']);
      final idx = props.map(allAppearanceProps.indexOf).toList();
      expect(idx, equals(List<int>.of(idx)..sort()), reason: 'orden estable = allAppearanceProps');
    });

    // Ratchet (KRO-133): TODA prop del catálogo debe gatearla al menos un kind.
    // Falla si `appearancePropsByKind` queda incompleto vs `allAppearanceProps`
    // (el render aplica las 26; la metadata del editor debe gatearlas todas).
    test('appearancePropsByKind cubre TODO el catálogo (sin props huérfanas)', () {
      final gated = <String>{for (final list in appearancePropsByKind.values) ...list};
      final missing = allAppearanceProps.where((p) => !gated.contains(p)).toList();
      expect(missing, isEmpty, reason: 'props del catálogo que ningún kind gatea: ${missing.join(', ')}');
    });
  });
}
